cpp/include/raft/spatial/knn/detail/ann_utils.cuh

```diff
@@ -573,4 +585,52 @@ struct batch_load_iterator {
   size_type cur_pos_;
 };
 
+template <typename IdxT>
+auto get_subsample_indices(raft::resources const& res, IdxT n_samples, IdxT n_subsamples, int seed)
+  -> raft::device_vector<IdxT, IdxT>
+{
+  RAFT_EXPECTS(n_subsamples <= n_samples, "Cannot have more training samples than dataset vectors");
+
+  auto data_indices = raft::make_device_vector<IdxT, IdxT>(res, n_samples);
+  thrust::counting_iterator<IdxT> first(0);
+  thrust::device_ptr<IdxT> ptr(data_indices.data_handle());
+  thrust::copy(raft::resource::get_thrust_policy(res), first, first + n_samples, ptr);
+  raft::random::RngState rng(seed);
+  auto train_indices = raft::make_device_vector<IdxT, IdxT>(res, n_subsamples);
+  raft::random::sample_without_replacement(res,
+                                           rng,
+                                           raft::make_const_mdspan(data_indices.view()),
+                                           std::nullopt,
+                                           train_indices.view(),
+                                           std::nullopt);
+  return train_indices;
+}
+
+/** Subsample the dataset to create a training set*/
+template <typename T, typename IdxT = int64_t>
+void subsample(raft::resources const& res,
+               const T* input,
+               IdxT n_samples,
+               raft::device_matrix_view<T, IdxT> output,
+               int seed)
+{
+  IdxT n_dim   = output.extent(1);
+  IdxT n_train = output.extent(0);
+
+  raft::device_vector<IdxT, IdxT> train_indices =
+    get_subsample_indices<IdxT>(res, n_samples, n_train, seed);
+
+  cudaPointerAttributes attr;
+  RAFT_CUDA_TRY(cudaPointerGetAttributes(&attr, input));
+  T* ptr = reinterpret_cast<T*>(attr.devicePointer);
+  if (ptr != nullptr) {
```

Comment:
This looks like a good place to use [`raft::mdbuffer`](https://github.com/rapidsai/raft/blob/d7eebc68a2582d8e1ca699df3816091d0224dcf7/cpp/include/raft/core/mdbuffer.cuh#L394)

Reply:
It is actually not (or I misunderstand the suggestion). With `mdbuffer` we could make a copy and call a single (e.g. device) implementation of `gather`. What we want is a dispatch to use the new `gather` call, so that H2D copies are overlapped with gathering the data.

Your are still right that this is not the right place to dispatch. Ideally` ivf_pq::build` would have an `mdspan` input, and we just pass that along, and the dispatch would happen in `matrix::gather`. Since I have some questions how to do that right, I propose to discuss that in https://github.com/rapidsai/raft/issues/2111.